include/gridtools/common/int_vector.hpp

```diff
@@ -0,0 +1,161 @@
+
+        template <class I>
+        struct is_integral_zero<I,
+            std::enable_if_t<is_integral_constant<meta::second<I>>::value && meta::second<I>::value == 0>>
+            : std::true_type {};
+
+        template <class Key>
+        struct at_key_f {
+            template <class T>
+            GT_FUNCTION GT_CONSTEXPR decltype(auto) operator()(T const &arg) const {
+                return host_device::at_key<Key>(arg);
+            }
+        };
+
+        /**
+         * @brief Returns `int_vector` with elements removed that are `integral_constant<T, 0>`
+         */
+        template <class Vec>
+        GT_FUNCTION GT_CONSTEXPR auto normalize(Vec &&vec) {
+            using filtered_map_t = meta::filter<meta::not_<is_integral_zero>::apply, hymap::to_meta_map<Vec>>;
+            using keys_t = meta::transform<meta::first, filtered_map_t>;
+            using generators = meta::transform<at_key_f, keys_t>;
+            return tuple_util::host_device::generate<generators, hymap::from_meta_map<filtered_map_t>>(
+                std::forward<Vec>(vec));
+        }
+    } // namespace int_vector_impl_
+
+    template <class T>
+    struct is_int_vector : std::true_type {
+    }; // TODO proper type trait: is_hymap and (is_integral or is_gr_integral_constant)
+
+    template <class T>
+    inline constexpr bool is_int_vector_v = is_int_vector<T>::value;
+
+    namespace int_vector {
+        using int_vector_impl_::multiply;
+        using int_vector_impl_::normalize;
+        using int_vector_impl_::plus;
+
+        namespace arithmetic {
+            template <class Vec,
+                class Scalar,
+                std::enable_if_t<is_int_vector_v<Vec> &&
+                                     (std::is_integral_v<Scalar> || is_integral_constant<Scalar>::value),
+                    bool> = true>
+            GT_FUNCTION GT_CONSTEXPR auto operator*(Vec &&vec, Scalar scalar) {
+                return multiply(std::forward<Vec>(vec), scalar);
+            }
+
+            template <class Vec,
+                class Scalar,
+                std::enable_if_t<is_int_vector_v<Vec> &&
+                                     (std::is_integral_v<Scalar> || is_integral_constant<Scalar>::value),
+                    bool> = true>
+            GT_FUNCTION GT_CONSTEXPR auto operator*(Scalar &&scalar, Vec vec) {
+                return multiply(std::forward<Vec>(vec), scalar);
+            }
+
+            template <class First,
+                class Second,
+                std::enable_if_t<is_int_vector_v<First> && is_int_vector_v<Second>, bool> = true>
+            GT_FUNCTION GT_CONSTEXPR auto operator+(First &&first, Second &&second) {
+                return plus(std::forward<First>(first), std::forward<Second>(second));
+            }
+
+            template <class Vec, std::enable_if_t<is_int_vector_v<Vec>, bool> = true>
+            GT_FUNCTION GT_CONSTEXPR auto operator-(Vec &&vec) {
```

Comment:
probably unary `operator+` should be here for  consistency :)

Reply:
done